ionoscloud/data_source_dataplatform_cluster.go

```diff
@@ -0,0 +1,373 @@
+							Computed: true,
+							Elem: &schema.Resource{
+								Schema: map[string]*schema.Schema{
+									"name": {
+										Type:      schema.TypeString,
+										Computed:  true,
+										Sensitive: true,
+									},
+									"context": {
+										Type:      schema.TypeMap,
+										Computed:  true,
+										Sensitive: true,
+										Elem: &schema.Schema{
+											Type: schema.TypeString,
+										},
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			"user_tokens": {
+				Type:      schema.TypeMap,
+				Sensitive: true,
+				Computed:  true,
+				Elem: &schema.Schema{
+					Type:      schema.TypeString,
+					Sensitive: true,
+				},
+			},
+			"ca_crt": {
+				Type:      schema.TypeString,
+				Sensitive: true,
+				Computed:  true,
+			},
+			"server": {
+				Type:      schema.TypeString,
+				Sensitive: true,
+				Computed:  true,
+			},
+			"kube_config": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+		},
+		Timeouts: &resourceDefaultTimeouts,
+	}
+}
+
+func dataSourceDataplatformClusterRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(SdkBundle).DataplatformClient
+
+	idValue, idOk := d.GetOk("id")
+	nameValue, nameOk := d.GetOk("name")
+
+	id := idValue.(string)
+	name := nameValue.(string)
+
+	if idOk && nameOk {
+		diags := diag.FromErr(errors.New("id and name cannot be both specified in the same time"))
+		return diags
+	}
+	if !idOk && !nameOk {
+		diags := diag.FromErr(errors.New("please provide either the Dataplatform Cluster id or name"))
+		return diags
+	}
+
+	var cluster dataplatform.ClusterResponseData
+	var err error
+
+	if idOk {
+		/* search by ID */
+		cluster, _, err = client.GetClusterById(ctx, id)
+		if err != nil {
+			diags := diag.FromErr(fmt.Errorf("an error occurred while fetching the Dataplatform Cluster with ID %s: %w", id, err))
+			return diags
+		}
+	} else {
+		var results []dataplatform.ClusterResponseData
+
+		partialMatch := d.Get("partial_match").(bool)
+
+		log.Printf("[INFO] Using data source for Dataplatform Cluster by name with partial_match %t and name: %s", partialMatch, name)
+
+		if partialMatch {
+			clusters, _, err := client.ListClusters(ctx, name)
+			if err != nil {
+				diags := diag.FromErr(fmt.Errorf("an error occurred while fetching Dataplatform Clusters: %w", err))
+				return diags
+			}
+			results = *clusters.Items
+		} else {
+			clusters, _, err := client.ListClusters(ctx, "")
+			if err != nil {
+				diags := diag.FromErr(fmt.Errorf("an error occurred while fetching Dataplatform Clusters: %w", err))
+				return diags
+			}
+			if clusters.Items != nil && len(*clusters.Items) > 0 {
+				for _, clusterItem := range *clusters.Items {
+					if clusterItem.Properties != nil && clusterItem.Properties.Name != nil && strings.EqualFold(*clusterItem.Properties.Name, name) {
+						tmpCluster, _, err := client.GetClusterById(ctx, *clusterItem.Id)
```

Comment:
@cristiGuranIonos first we make:

`clusters, _, err := client.ListClusters(ctx, "")`

Then we iterate through them make and make other requests for the clusters that match the name:

```
if clusterItem.Properties != nil && clusterItem.Properties.Name != nil && strings.EqualFold(*clusterItem.Properties.Name, name) {
						tmpCluster, _, err := client.GetClusterById(ctx, *clusterItem.Id)
```
Is this necessary? Can't we just iterate through previously retrieved `clusters` and build the list of results based on the name? Why do we need to make additional requests?

More than that, after building the list of `results`, we check if the length is greater than 1 and, if it is, we raise an error:
```
} else if len(results) > 1 {
			return diag.FromErr(fmt.Errorf("more than one Dataplatform Cluster found with the specified criteria name = %s", name))
}
```
I think that a better approach would be to raise an error in a loop, as soon as we observe that the length is  greater than 1. Basically we are making some useless requests because if the length is greater than 1, we still raise an error.

Reply:
done